signer/src/blockchain_sync.rs

```diff
@@ -0,0 +1,336 @@
+        // parameter. As we're walking backwards, we grab the first block in
+        // the tenure and use its parent block ID for the next iteration, which
+        // should give us the previous tenure and all of its blocks.
+        let first_block = tenure
+            .blocks()
+            .first()
+            .ok_or(Error::EmptyTenure(next_tenure_tip))?;
+
+        next_tenure_tip = first_block.header.parent_block_id;
+
+        // We re-use the same code as the block observer to write the blocks to
+        // the database.
+        block_observer::write_stacks_blocks(
+            &ctx.get_storage_mut(),
+            &ctx.config().signer.deployer,
+            &[tenure],
+        )
+        .await?;
+
+        // If the block height of the anchor block is equal to the nakamoto
+        // activation height then we're done.
+        if bitcoin_block.block_height == nakamoto_activation_height {
+            tracing::info!("nakamoto activation height reached, stacks block-sync completed");
+            return Ok(());
+        }
+    }
+}
+
+/// Helper function to wait for the Stacks node to reach the Nakamoto activation
+/// height.
+#[tracing::instrument(skip_all, fields(%nakamoto_activation_height))]
+pub async fn wait_for_nakamoto_activation_height(
+    ctx: &impl Context,
+    nakamoto_activation_height: u64,
+) -> Result<(), Error> {
+    let mut term = ctx.get_termination_handle();
+    let stacks_client = ctx.get_stacks_client();
+
+    let mut interval = tokio::time::interval(Duration::from_secs(5));
+    loop {
+        if let Ok(node_info) = stacks_client.get_node_info().await {
+            if node_info.burn_block_height > nakamoto_activation_height {
+                tracing::info!(
+                    current_height = %node_info.burn_block_height,
+                    "stacks node has reached the nakamoto activation height"
+                );
+
+                break;
+            }
+
+            tracing::info!(
+                current = %node_info.burn_block_height,
+                target = %nakamoto_activation_height,
+                "waiting for stacks node to reach nakamoto activation height"
+            );
+        }
+
+        tokio::select! {
+            _ = term.wait_for_shutdown() => return Ok(()),
+            _ = interval.tick() => {},
+        }
+    }
+
+    Ok(())
+}
+
+/// Waits for the Stacks node to report that it is fully-synced. It does this
+/// by polling the node's `info` endpoint every 5 seconds until the response's
+/// `is_fully_synced` field is `true`.
+#[tracing::instrument(skip_all)]
+pub async fn wait_for_stacks_node_to_report_full_sync(ctx: &impl Context) -> Result<(), Error> {
+    let mut term = ctx.get_termination_handle();
+    let stacks_client = ctx.get_stacks_client();
+
+    let mut interval = tokio::time::interval(Duration::from_secs(5));
+    loop {
+        match stacks_client.get_node_info().await {
+            Ok(node_info) if node_info.is_fully_synced => break,
```

Comment:
Given that in the nodes stalled on devnet this returns `true`, can we rely on this?

Reply:
Most likely not -- all of my assumptions about this flag have been thrown out the window 😬 We'll nees to ask what it _actually_ means... it could be that it's in sync with bitcoin, or it could mean that it's in sync with other nodes when it comes to receiving (but not processing) stacks blocks... or something completely different